src/main/java/io/github/thebusybiscuit/exoticgarden/PlantsListener.java

```diff
@@ -123,67 +186,102 @@ public void onGenerate(ChunkPopulateEvent e) {
-                                current.getRelative(BlockFace.UP).setBlockData(s);
-                                SkullBlock.setFromHash(current.getRelative(BlockFace.UP), berry.getTexture());
-                            });
-                            break;
-                        default:
+                if (PaperLib.isChunkGenerated(e.getWorld(), x, z)) {
+                    for (int y = e.getWorld().getMaxHeight(); y > 30; y--) {
+                        Block current = e.getWorld().getBlockAt(x, y, z);
+                        if (!current.getType().isSolid() && current.getType() != Material.WATER && berry.isSoil(current.getRelative(BlockFace.DOWN).getType())) {
+                            BlockStorage.store(current, berry.getItem());
+                            switch (berry.getType()) {
+                                case BUSH:
+                                    current.setType(Material.OAK_LEAVES);
+                                    break;
+                                case FRUIT:
+                                    current.setType(Material.PLAYER_HEAD);
+                                    Rotatable s = (Rotatable) current.getBlockData();
+                                    s.setRotation(faces[random.nextInt(faces.length)]);
+                                    current.setBlockData(s);
+                                    SkullBlock.setFromHash(current, berry.getTexture());
+                                    break;
+                                case ORE_PLANT:
+                                case DOUBLE_PLANT:
+                                    BlockStorage.store(current.getRelative(BlockFace.UP), berry.getItem());
+                                    current.setType(Material.OAK_LEAVES);
+                                    current.getRelative(BlockFace.UP).setType(Material.PLAYER_HEAD);
+                                    Rotatable ss = (Rotatable) current.getRelative(BlockFace.UP).getBlockData();
+                                    ss.setRotation(faces[random.nextInt(faces.length)]);
+                                    current.getRelative(BlockFace.UP).setBlockData(ss);
+                                    SkullBlock.setFromHash(current.getRelative(BlockFace.UP), berry.getTexture());
+                                    break;
+                                default:
+                                    break;
+                            }
                             break;
                         }
-                        break;
                     }
+                } else {
+                    PaperLib.getChunkAtAsync(e.getWorld(), x, z).thenRun(() -> {
+                        for (int y = e.getWorld().getMaxHeight(); y > 30; y--) {
+                            Block current = e.getWorld().getBlockAt(x, y, z);
+                            if (!current.getType().isSolid() && current.getType() != Material.WATER && berry.isSoil(current.getRelative(BlockFace.DOWN).getType())) {
```

Comment:
Both branches do the same, this redundancy can be massively reduced by making a seperate method for it.